### ISSUE TITLE
Incorrect scope handling for comprehension within class

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#2977]
Test case:

```python
class Capybara(object):
    incisors = [1, 2]
    canines = {incisors[0] for _ in incisors}
```

Pyright (via Pylance v2022.1.5) produces no errors for this code, but it fails at runtime:

```
% python 
Python 3.10.0rc1 (default, Aug 30 2021, 09:35:02) [Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class Capybara(object):
...     incisors = [1, 2]
...     canines = {incisors[0] for _ in incisors}
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in Capybara
  File "<stdin>", line 3, in <setcomp>
NameError: name 'incisors' is not defined
```

Note that the error is from the use of `incisors[0]` within the *body* of the comprehension; the use of `incisors` as the outermost iterable in the comprehension is fine. The outermost iterable is evaluated within the enclosing scope, but the body of the comprehension behaves like a function and therefore does not see the scope of the enclosing class.

**Comment by erictraut** [comment]
Thanks for reporting.

What makes this different from a comprehension evaluated within the global scope or from a function scope?

```python
incisors = [1, 2]
canines = {incisors[0] for _ in incisors}

def func():
    incisors = [1, 2]
    canines = {incisors[0] for _ in incisors}

func()
```

**Comment by erictraut** [comment]
This also works fine:
```python
incisors = [1, 2]
class Capybara(object):
    canines = {incisors[0] for _ in [1, 2, 3]}
```

**Comment by JelleZijlstra** [comment]
It's just how class scopes work: nested functions within the class don't get access to the class scope. Guido's in-progress scope blog has an explanation for why it is that way: https://gvanrossum.github.io/formal/scopesblog.html.

I'm not sure the interaction with comprehensions is really intentional, but that's how it works. PEP 572 (assignment expressions, https://www.python.org/dev/peps/pep-0572/#changing-the-scope-rules-for-comprehensions) considered changing the rules here.

Pyright implements this correctly within methods:
```python
class Capybara(object):
    incisors = [1, 2]
    canines = {incisors[0] for i in incisors}

    def f(self):
        print(incisors)  # error, incisors is undefined
```

**Comment by JelleZijlstra** [comment]
I found another case where pyright gets it wrong:
```python
class Capybara(object):
    incisors = [1, 2]
    x = lambda: incisors
    x()
```
This also fails at runtime but pyright gives no errors. Lambdas follow the same scoping rules as normal functions here, and as far as I'm aware always. (But Guido has a separate class for lambda scopes, so maybe I'm missing something.)

**Comment by erictraut** [comment]
Both cases (the list comprehension and the lambda) will be fixed in the next release.

**Comment by JelleZijlstra** [comment]
Thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.218, which I just published. It will also be included in the next release of pylance.
